### PR TITLE
Plumbing [Not dead, I'm just burnt out] [nevermind, definitely dead]

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Sinks, toilets, showers, and other water-reliant objects now draw from a plumbing tank that generates reagents for the entire station. These reagents can be changed by anyone who can access the tank.
/:cl:

WIP. I put this out because I didn't want to finish the entire feature only to have it be closed, and it's actually somewhat functional as of now.

Currently the sprite is a placeholder. If you feel like making a cool sprite I'd love to go with it.

Adds a new machine, the master plumbing tank. ___I haven't decided which job will get this yet, but I'm leaning towards engineering or atmos. I don't want to give it to janitors, because that is a job that should, in my opinion, stay shallow and not require much work.___ The tank automatically generates reagents up to a maximum of 10,000 units based on an internal "buffer" container. This is done at a rate of 1u a tick but will likely be increased if its parts are upgraded. By default, the buffer is a 30u bottle of water.

The tank has three modes: Idle, Synthesize, and Drain. Idle does just that and nothing more. Synthesize uses power to create more reagents, and Drain uses power to destroy them. Modes and all information can be read from the display, which, being a giant piece of plumbing, is difficult to understand, although there'll be a paper with helpful* notes next to it once I decide where to slap it down.

Also refactors all water-based machinery and structures to, unless overriden, draw from the master plumbing tank through bluespace magic. It isn't picky as to what it grabs, meaning that you can fill the pipes with acid and the next person to wash their hands will either lose their gloves or have their skin melt off.

I made this because it's something people have wanted for a long time and it's a fun addition to the game. I also operate under the assumption that I'll eventually be giving this to Engineering, so it gives them something to do other than shut off random APCs for amusement and micromanage the supermatter.

Also adds several helper procs for reagent holders to get the proportions of every reagent and to adjust them based on provided volume.

### CODE REVIEW

**Comment by optimumtact** [comment]
This isn't very mechanically interesting, can you make the machinery more complex or interesting for engineers/atmos to use? I.e require plasma or freon or similar

**Comment by Xhuis** [comment]
Having it require gases is just tedious and has no depth to it (unless you hook it up to a gas line, of course), but I can easily see about adding different things, either in the form of required maintenance or just better output if you fine-tune it enough. Things like that.

**Comment by optimumtact** [comment]
The supermatter does things right, every other machine is pretty much garbage.

**Comment by OtherDalfite** [comment]
This is a cool idea, but I don't really think it offers a lot to the game. If people wanna fuck over someone using acid they'll just use a spray bottle, and nobody really washes there hands anyways. I can only see this being used to like screw over botany or medbay, not much else.

**Comment by Xhuis** [comment]
@OtherDalfite Well, excuse me! I wash my hands in-game all the time.

Joking aside, it's not the most deep of features, but it's a little niche that people have wanted for a long time, and it's something I've always wanted to do. It can also be used in good and creative ways - like, if fires break out, engineers can make some kelotane and pump it into the pipes so that people can drink from the taps and heal their burn wounds. Overall, it's not extraordinarily deep, but it has a lot of uses, and not all of them have to be bad ones.

**Comment by Niknakflak** [comment]
Give it to the Janitor. Engis/atmos has enough toys.  If you seriously give plumbing to anyone but the fucking janitor, what the fuck are you doing?

**Comment by JStheguy** [comment]
No, you are all stupid.
Give it to the bartender, obviously.

**Comment by Xhuis** [comment]
@Niknakflak While the janitor is the very obvious logical choice, it has its appeal to most people (myself included) because it's a low-stress job that can still contribute to the station. I don't want to force a job to do something when it's renowned for being laid back.

**Comment by Xhuis** [comment]
Sinks now work under the new plumbing system. Code is a little spaghetti because of subtypes that don't use the new system, like puddles, but it's tested and seems to work okay.

**Comment by Jordie0608** [comment]
@Xhuis Asis this sounds like it could be comformtably ignored by a Janitor if he didn't want to get involved with it so I don't think it'd jeopardize a janitor's laid-back approach.

**Comment by Shapsy** [comment]
Do janitors work on plumbing? Sounds more mechanical than washing floors

**Comment by Xhuis** [comment]
Still not sure about that part.

**Comment by bgobandit** [comment]
HAPPENING

**Comment by PKPenguin321** [comment]
j a n i t o r
then give him a red hat and overalls as an alt outfit
PLEASE

**Comment by Niknakflak** [comment]
Janitors do plumbing
Plumbers do plumbing
Mechanical engineers do not do plumbing
Atmospheric engineers do not do plumbing

**Comment by GunHog** [comment]
I second the plumber = janitor idea, strictly because the number of players naming themselves "Mario" will increase ten-fold! On a serious note, it should not hurt the janitor's lazy, laid back lifestyle. He does not have to touch the machine unless someone breaks in and pours acid/fire/booze into it.

**Comment by Poojawa** [comment]
Atmosia is the city of pipes. They handle all aspects of the station's environment, including water/sewage handling in addition to air. 

Also Atmosia has enough space once that offset pipes PR is put in.

**Comment by Iamgoofball** [comment]
give it to the janitor

**Comment by optimumtact** [comment]
@Poojawa that would only make sense if this actually had any pipes associated with it.

**Comment by Poojawa** [comment]
Wasn't it assumed already that the pipes were already linked to all the sinks/toliets, but they would otherwise be too much of a pain to have to manage from a coding aspect?

**Comment by Xhuis** [comment]
Yeah, currently there are no plumbing pipes and it operates through code magic. I'll come up with fluff for it but I don't want to use pipes for it. I will be adding construction for sinks and whatnot that requires pipes, but they won't be linked physically to the tank.

There's still some division over the role that should get the tank, so I made a poll. Place ya bets.

**Comment by Xhuis** [comment]
Fucking phone.
http://www.strawpoll.me/12575071

**Comment by ShadowDeath6** [comment]
wasn't there another "plumbing" pr that got disliked because it was infinite duplication of meme chems?
I can't find it but I swear there was one

**Comment by Supermichael777** [comment]
clearly this is a job for the Chaplin

**Comment by KorPhaeron** [comment]
It doesn't matter what job you give it to unless you give it to chemistry because they are the only one with access to liquids that would meaningfully impact this.

Other than yelling at botany to bring one of their 10k unit tanks so they can fill the sinks roundstart I guess

File: code/modules/plumbing/_the_tank.dm [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			use_power = 2
+			reagents.add_reagent_list(buffer.reagents.get_reagent_proportions())
+		if(PLUMBING_TANK_DRAIN)
+			use_power = 2
+			for(var/reagent in reagents.reagent_list)
+				var/datum/reagent/R = reagent
+				reagents.remove_reagent(R.id, 1)
+
+/obj/machinery/plumbing_tank/interact(mob/user)
+	var/dat
```
Comment (RemieRichards) [inline code/modules/plumbing/_the_tank.dm]: ```var/list/dat = list()```

File: code/modules/plumbing/_the_tank.dm [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			if(PLUMBING_TANK_IDLE)
+				dat += "<font color='#003FFF'>An LED labeled \"OPMD\" is glowing a steady blue.</font>"
+			if(PLUMBING_TANK_SYNTHESIZE)
+				dat += "<font color='#00FF00'>An LED labeled \"OPMD\" is flickering green[reagents.total_volume == reagents.maximum_volume ? "" : ", and you hear flowing liquid"].</font>"
+			if(PLUMBING_TANK_DRAIN)
+				dat += "<font color='#FF0000'>An LED labeled \"OPMD\" is blinking red, and you hear [reagents.total_volume ? "the sound of draining liquid" : "sucking sounds"].</font>"
+	dat += "<br><br>"
+	dat += "<a href='?src=\ref[src];switch_modes=1'>Set the OPMD dial to a new setting</a>"
+	var/datum/browser/popup = new(user, "plumbing_tank", name, 600, 400)
+	popup.set_content(dat)
```
Comment (RemieRichards) [inline code/modules/plumbing/_the_tank.dm]: ```dat.Join()```

File: code/modules/plumbing/_the_tank.dm [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	var/mob/living/carbon/user = usr
+	if(href_list["switch_modes"])
+		var/new_mode = alert(user, "There are three options for the dial.", "OPMD Dial", "IDLE", "SYNT", "DRAI")
+		var/mode_fluff = ""
+		if(!user.canUseTopic(src) || !src || QDELETED(src) || stat)
+			return
+		switch(new_mode)
+			if("IDLE")
+				operating_mode = PLUMBING_TANK_IDLE
+			if("SYNT")
```
Comment (RemieRichards) [inline code/modules/plumbing/_the_tank.dm]: just add an extra letter, it makes SYNTH and DRAIN correct, and IDLE already is.

File: code/modules/plumbing/sink.dm [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	..()
+
+/obj/structure/sink/attack_hand(mob/living/user)
+	if(!user || !istype(user))
+		return
+	if(!iscarbon(user))
+		return
+	if(!Adjacent(user))
+		return
+	if(!reagents.total_volume && !plumbing_has_reagents(15) && use_reagents) //15 units per use
```
Comment (RemieRichards) [inline code/modules/plumbing/sink.dm]: put ```use_reagents``` first.

File: code/modules/plumbing/sink.dm [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+		return
+	if(use_reagents)
+		master_plumber.request_liquid(src, 15)
+	var/selected_area = parse_zone(user.zone_selected)
+	var/washing_face = 0
+	if(selected_area in list("head", "mouth", "eyes"))
+		washing_face = 1
+	user.visible_message("<span class='notice'>[user] starts washing their [washing_face ? "face" : "hands"]...</span>", \
+						"<span class='notice'>You start washing your [washing_face ? "face" : "hands"]...</span>")
+	busy = 1
```
Comment (RemieRichards) [inline code/modules/plumbing/sink.dm]: proper ```TRUE/FALSE```, come on.

File: code/modules/plumbing/sink.dm [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+	if(!do_after(user, 40, target = src))
+		busy = 0
+		return
+
+	busy = 0
+
+	if(!reagents.total_volume && !plumbing_has_reagents(15) && use_reagents)
```
Comment (RemieRichards) [inline code/modules/plumbing/sink.dm]: ```use_reagents``` first.

File: code/modules/plumbing/sink.dm [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	if(istype(O, /obj/item/stack/medical/gauze))
+		var/obj/item/stack/medical/gauze/G = O
+		new /obj/item/weapon/reagent_containers/glass/rag(src.loc)
+		to_chat(user, "<span class='notice'>You tear off a strip of gauze and make a rag.</span>")
+		G.use(1)
+		return
+
+	if(!istype(O))
+		return
+	if(O.flags & ABSTRACT) //Abstract items like grabs won't wash. No-drop items will though because it's still technically an item in your hand.
```
Comment (RemieRichards) [inline code/modules/plumbing/sink.dm]: combine these two ifs.

File: code/modules/plumbing/sink.dm [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+		busy = 0
+		if(use_reagents)
+			master_plumber.request_liquid(src, 15)
+		O.clean_blood()
+		O.acid_level = 0
+		reagents.reaction(O, TOUCH)
+		reagents.clear_reagents()
+		user.visible_message("<span class='notice'>[user] washes [O] using [src].</span>", \
+							"<span class='notice'>You wash [O] using [src].</span>")
+		return 1
```
Comment (RemieRichards) [inline code/modules/plumbing/sink.dm]: ```TRUE```

File: code/modules/reagents/chemistry/holder.dm [hunk 9]
```diff
@@ -743,6 +743,24 @@ var/const/INJECT = 5 //injection
 
 	return english_list(out, "something indescribable")
 
+/datum/reagents/proc/get_reagent_proportions(include_empty_space = FALSE) //Returns an assoclist with all reagents equal to the percentage of the container they're taking up
+	var/list/reagent_props = list()
+	if(!include_empty_space)
+		for(var/datum/reagent/reagent in reagent_list)
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/holder.dm]: istypeless please.

**Comment by Kraseo** [comment]
now i can finally fluoridate the station's water supply

**Comment by optimumtact** [comment]
FLOURIDATION CAUSES CANCER SCREEE

**Comment by factoryman942** [comment]
>be cmo
>draw 5u omnizine from your hypospray into  a bottle
>infinite omnizine

**Comment by optimumtact** [comment]
also yeah this is basically a massive duplication system for chems, guess I'm 👎  on it

**Comment by Kraseo** [comment]
omnizine is trash

**Comment by RemieRichards** [comment]
could make the machine use a godawful amount of power when set to anything not water.

**Comment by FantasticFwoosh** [comment]
@KorPhaeron giving to chemistry is a terrible idea because they or somebody in medical would pollute the water system readily every round with flourine or some other dangerous meme chemical (virologist drops entire bottle of infection diseases like legionnaires down the drain) 

> Right now the possibilities are scary, especially with xenobio chemicals, you'd have to constantly test the taps.

[I already had my own suggestion on who you could give it to,](https://tgstation13.org/phpBB/viewtopic.php?f=9&t=10257&p=271016#p270233) given that engineering should focus on the supermatter not going nuclear on them/maintenance and atmos usually steps in to help engineering and has their own semi complicated air system to handle. Besides the meme content of my suggestion its all very rational

> Emagging the water machine to produce sickness/toxicity and plant harming sewage

File: code/modules/plumbing/_the_tank.dm (hunk 1 above)
Comment (Xhuis) [inline code/modules/plumbing/_the_tank.dm]: Not once have I ever seen dat as a list. Why use it now?

File: code/modules/plumbing/_the_tank.dm (hunk 3 above)
Comment (Xhuis) [inline code/modules/plumbing/_the_tank.dm]: They're restricted to four letters for fluff purposes. The display is meant to be obscure.

File: code/modules/plumbing/sink.dm (hunk 5 above)
Comment (Xhuis) [inline code/modules/plumbing/sink.dm]: This is legacy code that was moved to a new file. It'll be touched up after I migrate the other objects.

**Comment by Xhuis** [comment]
The potential for reagent duplication is somewhat circumvented by the staggering amount of water already in the system meaning that you have to spend ten minutes draining the entire system. The Odysseus does the same thing already and at a much faster rate. That said, I'm going to blacklist certain reagents and will add some sort of cost to duplicating non-whitelisted reagents.

**Comment by FantasticFwoosh** [comment]
We have biohazard disposal i guess for just emergency dumping water, won't water dissapear when it touches space?

**Comment by Xhuis** [comment]
I don't know his real-life physics dictate what happens in that case, but for the purposes of this PR it can be arranged.

**Comment by kevinz000** [comment]
>i can't autism with pumping pipes

*glares*
did you implement liquid spills yet because I want to fill a room with water and see who tries going into it :^)

**Comment by FantasticFwoosh** [comment]
@PKPenguin321 give the jantior a green outfit and the plumber a red one.

Put the second job slot to have the spare yellow & purple outfits

File: code/modules/plumbing/_the_tank.dm (hunk 1 above)
Comment (RemieRichards) [inline code/modules/plumbing/_the_tank.dm]: >not once

Wow, well you've clearly never looked at a string UI in a while have you?
>Why use it now?

Relatively new feature (past year or so) is why. ```/list/Join()``` turns a list of strings into one string, simple enough.
Say I have this string: "C"
Then I ```+= "A"``` I now have "CA"
Then I ```+= "T"```` I now have "CAT"

The overall difference is "C" -> "CAT", Ok, makes sense.
But what has the string table done in the background?
NEW STRING "C"
NEW STRING "CA"
NEW STRING "CAT"
These strings will sit in the table, for a decent amount of time, in the hopes that they'll be used again, the issue is, NOBODY IS GOING TO USE "CA", "CA" SHOULD NOT BE IN THE TABLE, the *perfect* outcome would be:
NEW STRING "CAT"
("NEW STRING" here means "look through every single string and see if one of them matches, if so reuse it, else find an open spot and add the new one" spoilers: this isn't cheap when you're doing it every other line)

and by using a list, we get N strings (where N is the number of ```+=```'s) + 1 for the final string, now this *isn't* the perfect case, but it's a shit load better.

```/list/Join()``` was requested specifically for these kinds of situations, other languages with string tables/trees include a ```StringBuilder``` class, that holds onto each of the component strings, and produces one final string when you ```StringBuilder.finalize()```, our list + ```.Join()``` fulfils this role in DM.

tl;dr performance you dingbat

Comment (GunHog) [inline code/modules/plumbing/_the_tank.dm]: Wow, I did not know of this either, lol!

Comment (xxalpha) [inline code/modules/plumbing/_the_tank.dm]: This is basic beyond mechanics, guys. It's like you don't go to coderbus.

Comment (Xhuis) [inline code/modules/plumbing/_the_tank.dm]: I'm in there every day. I just added a console that used the old method and nobody batted an eye. Forgive my skepticism.

Comment (GunHog) [inline code/modules/plumbing/_the_tank.dm]: Maybe I was not in the chat at the time this standard was made, same with Xhuis! Most of us which missed that discussion will only find out when RR comes to correct us!

Comment (RemieRichards) [inline code/modules/plumbing/_the_tank.dm]: There was no discussion really, but I've enforced this for quite literally as long as the feature has existed (year, year+)

Comment (xxalpha) [inline code/modules/plumbing/_the_tank.dm]: It's been known for a good while now, but I am just messing with ya.

**Comment by KorPhaeron** [comment]
>The potential for reagent duplication is somewhat circumvented

Isn't that the sole feature of this though?

I can see this being used for:

1) Healing chems for the whole station

2) People making holocaust jokes in the showers

Both this system and our chemicals in general are pretty limited so there doesn't seem to be much room to do anything else

**Comment by Xhuis** [comment]
...I can't think of an argument against that. Yeah, you have a point.

**Comment by Xhuis** [comment]
In other news, @Nienhaus has graciously provided me sprites, so I'll be using those and posting a picture soon.

**Comment by Anonmare** [comment]
In regards to who should get the tank. I say put it near atmos and give both janitors and atmostechs access to it. Should be a decent compromise, no?

Also I hope this will allow for the potential of slippery-sloping into implementing liquid flooding one day, we are still playing an overdue signed atmospherics simulator after all. That and I like the idea of locking someone in the bathroom while it floods.

**Comment by Iatots** [comment]
Really not a fan of another magic bluespace feature, though even a real implementation with piping would be a nightmare with assistants unwrenching water pipes to flood the entire station.

**Comment by Jalleo** [comment]
@optimumtact people will probably abuse for two weeks realise its weak as fuck and then almost never use it.

**Comment by TrustyGun** [comment]
i dont see why you shouldnt give it to the janitor, its not like this feature is super complicated and crazy

**Comment by AdamElTablawy** [comment]
If you want it to have more impact than a simple niche thing, you'll want to add piping rather than bluespess magic, tie a few machines to it  and probably add water itself flowing and flooding.

For example, cryogenics AND the cloning pods both seem to have the occupant immersed in some kind of fluid, and I doubt that a tiny beaker can cover an entire human and more as the cryo tank seems to indicate. So, cryogenics fluid uses plumbing for water. Sabotage cryo? You don't just need to modify the beaker-You can dick with plumbing instead and suddenly cryo's fluid is fucking acid.

**Comment by Cyberboss** [comment]
With the planetstation stuff upcoming, design decisions about water have to be made eventually

**Comment by Xhuis** [comment]
I've been chatting with @AdamElTablawy and I'm aiming to expand this system a bunch in the future. Adding stuff like sprinklers to put out minor fires, adding liquid cooling to certain machines, that kind of thing.

**Comment by kevinz000** [comment]
On the note of pipe closers our pipes don't even leak if there's an opening so we don't need it :P

**Comment by Bawhoppen** [comment]
What is the point in plumbing without actual pipes???

**Comment by Xhuis** [comment]
What's the point of teleportation if you don't need to link them with wires?

I'd like pipes but it'd require a tremendous amount of work from mappers and players. I don't want _four_ stationwide lines.

File: code/modules/plumbing/sink.dm [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	user.visible_message("<span class='notice'>[user] washes their [washing_face ? "face" : "hands"] using [src].</span>", \
+						"<span class='notice'>You wash your [washing_face ? "face" : "hands"] using [src].</span>")
+	if(washing_face)
+		if(ishuman(user))
+			var/mob/living/carbon/human/H = user
+			H.lip_style = null //Washes off lipstick
+			H.lip_color = initial(H.lip_color)
+			H.wash_cream()
+			H.regenerate_icons()
+		user.drowsyness = max(user.drowsyness - rand(2,3), 0) //Washing your face wakes you up if you're falling asleep
```
Comment (Niknakflak) [inline code/modules/plumbing/sink.dm]: woah, EMERGENT GAMEPLAY

Comment (Xhuis) [inline code/modules/plumbing/sink.dm]: M E T A
E
T
A

File: code/modules/plumbing/sink.dm [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	qdel(src)
+
+
+
+/obj/structure/sink/kitchen
+	name = "kitchen sink"
+	icon_state = "sink_alt"
+
+
+/obj/structure/sink/puddle	//splishy splashy ^_^
```
Comment (Niknakflak) [inline code/modules/plumbing/sink.dm]: comment not required but definitely appreciated

Comment (Xhuis) [inline code/modules/plumbing/sink.dm]: This came from legacy code, I'm not _that_ autistic

File: code/modules/plumbing/sink.dm (hunk 10 above)
Comment (Iatots) [inline code/modules/plumbing/sink.dm]: showers should totally have this, especially when freezing.

**Comment by Xhuis** [comment]
DM is being finnicky and sometimes randomly never initializes the lighting subsystem. Testing on this probably won't happen quickly.

**Comment by Xhuis** [comment]
* Added damage, destruction, and repairing.
  * From its useless state, add five metal, weld, and wrench. Done!
* Added the ability to eject and add buffers.
* Code cleanup in response to maintainer review.

**Comment by FantasticFwoosh** [comment]
@KorPhaeron the problem with this is that you're forgetting its a WATER system and all the nooks and crannies it delivers WATER to the station else they use the water tanks (actually remove water tanks or use them as water canisters locally would be easier)

> There will be literal napalm showers, and phosphorus exploding sinks/pipes that punish people pouring water down a drain. You cant water plants with healing chems, the most annoying thing may be holy water for ruining a cultists day.

> Create canister pad (or water variant) with attached water pipes, to divert water in your 'gas form' to fuel the taps for current water/fuel containers meaning you can have all of botany's water supply in taps in the back room or maintenance without wiring it into the grid

> Also how will this new gas system affect Toxins? I have a suggestive thread up.

File: code/modules/plumbing/sink.dm (hunk 11 above)
Comment (Militaires) [inline code/modules/plumbing/sink.dm]: splishy splashy

**Comment by AdamElTablawy** [comment]
The issue with plumbing is-

Bombs become more unrepairable.

Water comes gushing out of broken pipes, so one busted pipe instantly dicks the water system and also makes breaches slip hell, which while amusing is still irritating.

That's 4 maps for plumbin'.

There should be drains, pumps, etc, but that's both out of the scope of this pr probs and also shouldn't be interconnected in that way.

**Comment by FantasticFwoosh** [comment]
A small localised pipe system linked to tanks in places such as starboard storage & maint could work for closed water supplies not attached to the mains. As i previously mentioned.

A bit like bringing a air canister with you to go with the aux base, putting in a input then tying up the output to connect to the 'disposal water' network to be shot out of biohazard disposal into space where it gets sucked into the void.

**Comment by Xhuis** [comment]
I should elaborate that the scope of this PR is mainly to migrate all existing water-based structures to the new system. That's the extent of it currently - other features will come in later PRs after people have had a chance to demo the system and if they like it.

File: code/modules/plumbing/shower.dm [hunk 12]
```diff
@@ -0,0 +1,223 @@
+/obj/machinery/shower
```
Comment (RemieRichards) [inline code/modules/plumbing/shower.dm]: good job, thanks for making the diff hard to read.

**Comment by Xhuis** [comment]
Showers are done, but keep in mind that a lot of is legacy code moved to a new file so it might not be up to scratch.

**Comment by Xhuis** [comment]
remietime

**Comment by RemieRichards** [comment]
you had that prepared!

File: code/modules/plumbing/shower.dm (hunk 12 above)
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: Sorry, I guess? The parts I changed are everything involving reagents. I'll go through and highlight the particulars in a few but something just came up so it'll take me ~20 minutes at the most, hopefully

**Comment by Xhuis** [comment]
I totally did.

**Comment by Cyberboss** [comment]
I am saving this pic forever omfg

**Comment by Xhuis** [comment]
Turns out 20 minutes was a vast underestimate. Give me a minute...

File: code/modules/plumbing/shower.dm [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+	density = 0
+	anchored = 1
+	use_power = 0
+	var/on = 0
+	var/obj/effect/mist/mymist = null
+	var/ismist = 0				//needs a var so we can make it linger~
+	var/watertemp = "normal"	//freezing, normal, or boiling
+
+/obj/machinery/shower/Initialize()
+	create_reagents(5)
```
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: This is new, obviously.

File: code/modules/plumbing/shower.dm [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+	name = "mist"
+	icon = 'icons/obj/watercloset.dmi'
+	icon_state = "mist"
+	layer = FLY_LAYER
+	anchored = 1
+	mouse_opacity = 0
+
+
+/obj/machinery/shower/attack_hand(mob/M)
+	if(!on && !plumbing_has_reagents(5) && !reagents.total_volume)
```
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: Showers can't function without water now.

File: code/modules/plumbing/shower.dm [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+		M << "<span class='warning'>Nothing comes out of the showerhead!</span>"
+		return
+	on = !on
+	update_icon()
+	add_fingerprint(M)
+	master_plumber.request_liquid(src, 5)
+	if(on)
+		process()
+
```
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: I also made this do what process() does because they do the same thing.

File: code/modules/plumbing/shower.dm [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+		var/turf/tile = loc
+		loc.clean_blood()
+		for(var/obj/effect/E in tile)
+			if(is_cleanable(E))
+				qdel(E)
+
+
+/obj/machinery/shower/proc/wash_mob(mob/living/L)
+	L.wash_cream()
+	if(iscarbon(L))
```
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: I cut the fire stack removal out of here, as I think it's what water does anyhow.

File: code/modules/plumbing/shower.dm [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+				reagents.chem_temp = 50
+			if("boiling")
+				reagents.chem_temp = 300
+			else
+				reagents.chem_temp = initial(reagents.chem_temp)
+		wash_turf()
+		reagents.reaction(loc, TOUCH)
+		for(var/atom/movable/G in loc)
+			reagents.reaction(G, TOUCH)
+			if(isliving(G))
```
Comment (Xhuis) [inline code/modules/plumbing/shower.dm]: Added the reagents reaction here. At the end of the line it clears all of them.

**Comment by Xhuis** [comment]
That should be all. The reagents are also heated or cooled to match if the water temperature is set to something.

**Comment by coiax** [comment]
@Xhuis I feel that you should probably remove the duplication aspect, and just start with a large water tank hooked up. I don't really like getting something "for nothing" as it were. I've got my eyes on cryo for that reason as well.

**Comment by FantasticFwoosh** [comment]
Its water for energy, else you have to do the annoying ordering water from cargo thing like we currently have to just buy more oxygen on the station in the rare instances we need it or atmos pipes are screwed.

If you want take up "Dirty deeds done dirt cheap" reasoning, go confront mekhi about his gas mining operations that magically make gas to replace virtually all the map gas cylinders. 

> (also destroy the water tank and enough water to flood the entire station 50 times comes rushing out and insta kills everything)

**Comment by coiax** [comment]
Ordering more water from cargo is exactly the sort of shit we should be encouraging.

**Comment by Xhuis** [comment]
once this is on the maps it should be ready for what this PR is aiming for. balance issues aside.

**Comment by kevinz000** [comment]
Honestly for the chemical stuff it should be poured in and only water and some basic reagents should be synthesized. Would be quite OP if some certain reagents got in there.

**Comment by Xhuis** [comment]
I could see that working but at that point you'd require massive amounts of reagents. Which I suppose would be an effective balancer, in all honestly.

**Comment by KorPhaeron** [comment]
@tgstation/commit-access 

This has been open 9 days now. If someone doesn't come along and declare they're willing to help this be finished and then hit the merge button within the next 48 hours I'm going to assume no maintainers want it and close it

**Comment by Xhuis** [comment]
You and 48 hour ultimatums, Kor. I don't think that that link even pings anyone as it leads to a 404 page. 

As for finishing this, its initial content is passable. I've just been burnt out lately and haven't been actively working on it, and I've been playing other games that don't sap my creative energies so much. It still needs balancing and putting on the map, so if I'm not up for working on it tomorrow I'll close it myself and reopen it once I've done work on it.

**Comment by RemieRichards** [comment]
@Xhuis it's a 404 for you, it's the list of contributors with commit access to those who have commit access (read: again, not you)

**Comment by Iamgoofball** [comment]
WHOA

LEAKING MAINTAINER SECRETS

**Comment by RemieRichards** [comment]
nothing about that is secret, you can work it all out from other sources.
oranges, Raz, MSO, Serverbot and Dorsis all have commit access but have their settings on "Private" in the team, so they don't show up, but everyone knows Dorsis tags things, Raz is a maintainer, oranges is, MSO has access because it's mso and the bot has it because it needs it for changelogs.

and the huge moby dick copypasta is... moby dick, a public book.

tl;dr shove off goofballs.

**Comment by Iamgoofball** [comment]
L
E
A
K
I
N
G

**Comment by AnturK** [comment]
It's okay it's not our world domination org projects.

As for the feature itself, i wouldn't mind it if it wasn't so effortless. I feel like it's healing sinks is going to become another "ugprade orm" must-have that happens every round because it's laughably easy.  And medbay made obsolete is  not something i'd like to see every round.

**Comment by RemieRichards** [comment]
@AnturK yeah, those are on the repositories tab, which isn't selected in my screenshot.

**Comment by AnturK** [comment]
They'll never know.

**Comment by optimumtact** [comment]
mine is usually public but I keep getting removed from the list

**Comment by lzimann** [comment]
@optimumtact Have you tried not commiting treason over and over again?

**Comment by coiax** [comment]
I don't want any synthesis of new reagents, I just want a giant tank that needs to be refilled.

**Comment by RemieRichards** [comment]
or that.

**Comment by lzimann** [comment]
About the PR: I don't think a giant tank that magically fills every sink with the reagent you put inside is fun in any way, it's basically no improvement from what it is today.

**Comment by GunHog** [comment]
>About the PR: I don't think a giant tank that magically fills every sink with the reagent you put inside is fun in any way, it's basically no improvement from what it is today.

Pouring clf3 into it to create FIRESHOWERS sounds like fun to me! Anyway, not having any sort of reagent creation or multiplication could mean that anything the janitor adds would affect at most only a few people. 

How else is anyone lacking access to a chem machine going to produce enough reagents to meaningfully affect people?

**Comment by Xhuis** [comment]
A few thoughts:

* A tank that has to be refilled can work, but it doesn't offer anything new besides a chore for someone to do. It's still possible to do though, and very easy.
* The whole point of reagent synthesis is to allow room for creativity. People might fill the sinks with burn medicine in the event of a fire to keep people alive, for instance. I think it _is_ an improvement over what we have, but it's admittedly a shallow one.
* If the feature really is "effortless", I'd be happy to make changes to it and try to make it more fun and interesting for everyone involved. While this is the framework, I'm still intending to add more at some point. At least, I would, if I wasn't burned out.

I'll have a verdict on whether or not I'll close this to continue later by this evening. Leaning towards that way though.

**Comment by KorPhaeron** [comment]
@gun_hog

It is popular because it sounds cool on the surface (fire showers!)

But in reality it just makes medbay even more redundant and worthless when you dump a beaker of omnizine or whatever in roundstart.

And it isn't worth throwing an already suffering department under the bus for the sake of fire showers.

Ironically, several of the people who upvoted this also post on the forums about wanting deeper medical.

**Comment by Xhuis** [comment]
One of its outstanding issues currently is that it lacks any balancing at all. You can just shove anything in there and get infinite amounts of it. Before the PR is merged I'd like to change the balancing in it. I'll probably make a whitelist of reagents that are cheap to produce, with the rest being slow and heavy on power drain. If some idiot tries to fill the pipes with deadly toxins, the power drain will be evident enough that an engineer can just shut off power remotely.

There are also other ways, and I'm open for suggestions.

**Comment by coiax** [comment]
@KorPhaeron I don't want to get too pedantic, but omnizine is not a chemical availible at roundstart. The best you could do with a generic roundstart cureall is tricoord.

**Comment by Xhuis** [comment]
Iirc you can grind warm donk pockets for omnizine. I used to do it for strange reagent roundstart as a chemist.

**Comment by coiax** [comment]
Oh

**Comment by KorPhaeron** [comment]
I think someone put adminrozine in the mining pens as well so if you're willing to do the grinding step you could have a magic nanite dispenser in every department within the first 5-10 minutes of the round

**Comment by coiax** [comment]
I thought they took the nanites out of the mining pens.

Anyway, I don't like magical synthesis of reagents in general, so I vote for giant tank of water.

**Comment by Xhuis** [comment]
I can't muster up the energy to work on this right now, and I'm not going to let it stay on the tracker for ages. I'll close this for now, and once I've made some progress open it up again.

**Comment by GunHog** [comment]
Good luck Xhuis! We are counting on you not to let this wind up abandoned!

**Comment by ShizCalev** [comment]
Suggestion; Give emagged chem dispensers have an option to dump contents of a beaker into the water line (since they distribute water, one would assume they were already connected to it)

**Comment by Xhuis** [comment]
@GunHog I didn't leave you hanging with shadowling, even if it did end up being shut down. I'm hoping this one won't go bad.

@ShizCalev Seems very strong to me but I may consider it.

**Comment by ShizCalev** [comment]
Will shuttles have to have a water tank added since some of them have showers/sinks too?

**Comment by Xhuis** [comment]
Probably, yeah.
